### PR TITLE
Detect more assertions in `prefer-t-regex`

### LINKED ISSUE #275: Detect more assertions in the `t-prefer-regex` rule
[Issuehunt badges](https://issuehunt.io/r/avajs/eslint-plugin-ava/issues/275)

See https://github.com/avajs/eslint-plugin-ava/issues/158#issuecomment-523830714

---

> *Note: This issue has a bounty, so it's expected that you are an experienced programmer and that you give it your best effort if you intend to tackle this. Don't forget, if applicable, to add tests, docs (double-check for typos). And don't be sloppy. Review your own diff multiple times and try to find ways to improve and simplify your code. Instead of asking too many questions, present solutions. The point of an issue bounty is to reduce my workload, not give me more. Include a 🦄 in your PR description to indicate that you've read this. Thanks for helping out 🙌* - @sindresorhus

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $40.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($40.00)

### Submitted pull Requests
- [#297 Add equality checks to `prefer-t-regex` rule](https://issuehunt.io/r/avajs/eslint-plugin-ava/pull/297)
- [#319 Detect more assertions in `prefer-t-regex`](https://issuehunt.io/r/avajs/eslint-plugin-ava/pull/319)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/avajs/eslint-plugin-ava/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

### PR DESCRIPTION
**Pull request by oOBoomberOo**
This pull request is a continuation from #297 work and fixes #275 

I've added support for `RegExp()` as requested in the previous PR as well as the necessary tests. 🦄

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#275: Detect more assertions in the `t-prefer-regex` rule](https://issuehunt.io/repos/49722492/issues/275)
---
</details>

### CODE REVIEW

File: test/prefer-t-regex.js [hunk 1]
```diff
@@ -61,6 +63,42 @@ ruleTester.run('prefer-t-regex', rule, {
...
+			errors: errors('notRegex')
+		},
+		{
+			code: header + 'test(t => t.true(new RegExp("\\d+").test(foo())));',
+			output: header + 'test(t => t.regex(foo(), new RegExp("\\d+")));',
+			errors: errors('regex')
+		},
+		{
+			code: header + 'const reg = RegExp("\\d+");\ntest(t => t.true(reg.test(foo.bar())));',
+			output: header + 'const reg = RegExp("\\d+");\ntest(t => t.regex(foo.bar(), reg));',
```
Comment (sindresorhus) [inline test/prefer-t-regex.js:100]: `RegExp` can also accept a regex literal. That needs to be tested as well: `RegExp(/foo/)`.

File: rules/prefer-t-regex.js [hunk 2]
```diff
@@ -13,75 +13,170 @@ const create = context => {
...
+		}
+
+		// Look up references in case it's a variable or RegExp declaration.
+		const reference = findRootReference(lookup);
+
+		if (reference) {
+			return reference.regex || reference.name === 'RegExp';
+		}
+
+		return false;
```
Comment (sindresorhus) [inline rules/prefer-t-regex.js:82]: Per code coverage warning, this is not covered by the tests.

File: rules/prefer-t-regex.js [hunk 3]
```diff
@@ -13,75 +13,183 @@ const create = context => {
...
-			const isBooleanAssertion = node.callee.type === 'MemberExpression' &&
-				booleanTests.has(node.callee.property.name) &&
-				util.getNameOfRootNodeObject(node.callee) === 't';
+		if (reference && reference.resolved) {
+			const definitions = reference.resolved.defs;
 
-			if (!isBooleanAssertion) {
-				return;
+			if (definitions.length === 0) {
+				return null;
```
Comment (sindresorhus) [inline rules/prefer-t-regex.js:29]:
```suggestion
				return;
```

I prefer not using `null` at all.

File: rules/prefer-t-regex.js [hunk 4]
```diff
@@ -13,75 +13,183 @@ const create = context => {
...
+			return definitions[definitions.length - 1].node;
+		}
+	};
 
-			// First argument is a call expression
-			const isFunctionCall = firstArg.type === 'CallExpression';
-			if (!isFunctionCall || !firstArg.callee.property) {
-				return;
+	// Recursively find the "origin" node of the given node.
+	// Note: Due to some limitation, this function will only find the reference up to two layer.
```
Comment (sindresorhus) [inline rules/prefer-t-regex.js:37]: What limitation? That should be documented here.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: rules/prefer-t-regex.js (hunk 3 above)
Comment (oOBoomberOo) [reply rules/prefer-t-regex.js]: Ah, I didn't know you have that preference. I've removed that now.

File: rules/prefer-t-regex.js (hunk 4 above)
Comment (oOBoomberOo) [reply rules/prefer-t-regex.js]: I believe the next commit should explain this thoroughly now.

**Pull request merged**